joblib/numpy_pickle_utils.py

```diff
@@ -0,0 +1,199 @@
+"""Utilities for fast persistence of big data, with optional compression."""
+
+# Author: Gael Varoquaux <gael dot varoquaux at normalesup dot org>
+# Copyright (c) 2009 Gael Varoquaux
+# License: BSD Style, 3 clauses.
+
+import pickle
+import sys
+import io
+
+PY3 = sys.version_info[0] >= 3
+PY26 = sys.version_info[0] == 2 and sys.version_info[1] == 6
+
+if PY3:
+    Unpickler = pickle._Unpickler
+    Pickler = pickle._Pickler
+    xrange = range
+else:
+    Unpickler = pickle.Unpickler
+    Pickler = pickle.Pickler
+
+
+def hex_str(an_int):
+    """Convert an int to an hexadecimal string."""
+    return '{0:#x}'.format(an_int)
+
+if PY3:
+    def asbytes(s):
+        if isinstance(s, bytes):
+            return s
+        return s.encode('latin1')
+else:
+    asbytes = str
+
+try:
+    import numpy as np
+except ImportError:
+    np = None
+
+# Compressed pickle header format: _ZFILE_PREFIX followed by
+# bytes which contains the length of the zlib compressed data as an
+# hexadecimal string. For example: 'ZF0x139              '
+_ZFILE_PREFIX = asbytes('ZF')
+_GZIP_PREFIX = b'\x1f\x8b'
+
+
+def gzip_file_factory(f, mode='rb', compresslevel=0):
+    """Factory to produce the class so that we can do a lazy import on gzip."""
+    import gzip
+    from gzip import WRITE
+
+    class GzipFile(gzip.GzipFile):
+
+        def _init_write(self, filename):
+            self.name = filename
+            self.crc = 0xffffffff
+            self.size = 0
+            self.writebuf = []
+            self.bufsize = 0
+
+        def write(self, data):
+            try:
+                # works with Python < 3.5
+                self._check_closed()
+            except AttributeError:
+                pass
+            if self.mode != WRITE:
+                import errno
+                raise OSError(errno.EBADF,
+                              "write() on read-only GzipFile object")
+
+            if self.fileobj is None:
+                raise ValueError("write() on closed GzipFile object")
+
+            # Convert data type if called by io.BufferedWriter.
+            if not PY26 and isinstance(data, memoryview):
+                data = data.tobytes()
```

Comment:
.tobytes is new in numpy 1.9 so I guess that means that there is no tests for this 'if clause'. Do we really need that?